src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpOnTypeFormattingPass.cs

```diff
@@ -84,6 +89,28 @@ public async override Task<FormattingResult> ExecuteAsync(FormattingContext cont
             return new FormattingResult(finalEdits);
         }
 
+        private TextEdit[] FilterCSharpTextEdits(FormattingContext context, TextEdit[] edits)
+        {
+            var filteredEdits = edits.Where(e => !AffectsWhitespaceInNonCSharpLine(e)).ToArray();
+            return filteredEdits;
+
+            bool AffectsWhitespaceInNonCSharpLine(TextEdit edit)
+            {
+                //
+                // Example:
+                //     @{
+                //       var x = "asdf";
+                // |  |}
+                // ^  ^ - C# formatter wants to remove this whitespace because it doesn't know about the '}'.
+                // But we can't let it happen.
+                //
+                return
+                    edit.Range.Start.Character == 0 &&
+                    edit.Range.Start.Line == edit.Range.End.Line &&
+                    !context.Indentations[(int)edit.Range.Start.Line].StartsInCSharpContext;
```

Comment:
Could you clarify this for me real quick. I figured `StartsInCSharpContext` meant that the line began in C# which I'd figure `___}` would qualify

Reply:
Open to better names but the logic is it looks for the first non-whitespace span in that line and returns its kind.

Reply:
So `___}` would be false as there's nothing but whitespace (in the generated document)?

Reply:
See [here](https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/IndentationContext.cs#L16). `FirstSpan` here refers to the first non-whitespace span

Reply:
I see I see. Ya I don't have much better idea for naming other then making it ungodly long lol.

Reply:
To clarify and to answer @TanayParikh's question, see [here](https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingContext.cs#L178). `FirstSpan` points to the span at position 0 if the line is all whitespace. So in our example, that will be CSharp context. So the edit will be allowed. Let me know if that is confusing still.

Reply:
Makes sense, thanks!